# Identifying key sales trends using SQL

In this project I will be analysing the database for the food distributor Northwind Traders and identifying key trends relating to employees, customers and products.

### Connecting to the database

In [1]:
%load_ext sql
%sql postgresql://postgres:5432@localhost/northwind

Connecting to 'postgresql://postgres:***@localhost/northwind'

In [2]:
#Removing limits on maximum number of rows displayed
%config SqlMagic.displaylimit = None

displaylimit: Value None will be treated as 0 (no limit)

### Database Tables

In [52]:
%%sql 
SELECT table_name,
       table_type 
  FROM information_schema.tables
 WHERE table_schema = 'public' AND table_type IN ('BASE TABLE');

Running query in 'postgresql://postgres:***@localhost/northwind'

14 rows affected.

+------------------------+------------+
|       table_name       | table_type |
+------------------------+------------+
|       suppliers        | BASE TABLE |
|        shippers        | BASE TABLE |
|         region         | BASE TABLE |
|       us_states        | BASE TABLE |
|       customers        | BASE TABLE |
|         orders         | BASE TABLE |
|       employees        | BASE TABLE |
|        products        | BASE TABLE |
|     order_details      | BASE TABLE |
|       categories       | BASE TABLE |
|      territories       | BASE TABLE |
|  employee_territories  | BASE TABLE |
| customer_demographics  | BASE TABLE |
| customer_customer_demo | BASE TABLE |
+------------------------+------------+

Our main focus will be on the tables relating to `customers`,`orders`, `employees`, `products`, `order_details` and `categories`. 

### Exploring the Data

Below displays the first 5 rows of some of the tables we will be working with.

In [17]:
%%sql
SELECT *
  FROM orders
 LIMIT 5;

Running query in 'postgresql://postgres:***@localhost/northwind'

5 rows affected.

+----------+-------------+-------------+------------+---------------+--------------+----------+---------+---------------------------+----------------------+----------------+-------------+------------------+--------------+
| order_id | customer_id | employee_id | order_date | required_date | shipped_date | ship_via | freight |         ship_name         |     ship_address     |   ship_city    | ship_region | ship_postal_code | ship_country |
+----------+-------------+-------------+------------+---------------+--------------+----------+---------+---------------------------+----------------------+----------------+-------------+------------------+--------------+
|  10248   |    VINET    |      5      | 1996-07-04 |   1996-08-01  |  1996-07-16  |    3     |  32.38  | Vins et alcools Chevalier |  59 rue de l'Abbaye  |     Reims      |     None    |      51100       |    France    |
|  10249   |    TOMSP    |      6      | 1996-07-05 |   1996-08-16  |  1996-07-10  |    1     |  11.61  |     Toms Spezialitäten    |    Luisenstr. 48     |    Münster     |     None    |      44087       |   Germany    |
|  10250   |    HANAR    |      4      | 1996-07-08 |   1996-08-05  |  1996-07-12  |    2     |  65.83  |       Hanari Carnes       |   Rua do Paço, 67    | Rio de Janeiro |      RJ     |    05454-876     |    Brazil    |
|  10251   |    VICTE    |      3      | 1996-07-08 |   1996-08-05  |  1996-07-15  |    1     |  41.34  |    Victuailles en stock   |  2, rue du Commerce  |      Lyon      |     None    |      69004       |    France    |
|  10252   |    SUPRD    |      4      | 1996-07-09 |   1996-08-06  |  1996-07-11  |    2     |   51.3  |      Suprêmes délices     | Boulevard Tirou, 255 |   Charleroi    |     None    |      B-6000      |   Belgium    |
+----------+-------------+-------------+------------+---------------+--------------+----------+---------+---------------------------+----------------------+----------------+-------------+------------------+--------------+

In [16]:
%%sql
SELECT *
  FROM order_details
 LIMIT 5;

Running query in 'postgresql://postgres:***@localhost/northwind'

5 rows affected.

+----------+------------+------------+----------+----------+
| order_id | product_id | unit_price | quantity | discount |
+----------+------------+------------+----------+----------+
|  10248   |     11     |    14.0    |    12    |   0.0    |
|  10248   |     42     |    9.8     |    10    |   0.0    |
|  10248   |     72     |    34.8    |    5     |   0.0    |
|  10249   |     14     |    18.6    |    9     |   0.0    |
|  10249   |     51     |    42.4    |    40    |   0.0    |
+----------+------------+------------+----------+----------+

In [39]:
%%sql
SELECT *
  FROM customers
 LIMIT 5;

Running query in 'postgresql://postgres:***@localhost/northwind'

5 rows affected.

+-------------+------------------------------------+--------------------+----------------------+-------------------------------+-------------+--------+-------------+---------+----------------+----------------+
| customer_id |            company_name            |    contact_name    |    contact_title     |            address            |     city    | region | postal_code | country |     phone      |      fax       |
+-------------+------------------------------------+--------------------+----------------------+-------------------------------+-------------+--------+-------------+---------+----------------+----------------+
|    ALFKI    |        Alfreds Futterkiste         |    Maria Anders    | Sales Representative |         Obere Str. 57         |    Berlin   |  None  |    12209    | Germany |  030-0074321   |  030-0076545   |
|    ANATR    | Ana Trujillo Emparedados y helados |    Ana Trujillo    |        Owner         | Avda. de la Constitución 2222 | México D.F. |  None  |    05021    |  Mexico |  (5) 555-4729  |  (5) 555-3745  |
|    ANTON    |      Antonio Moreno Taquería       |   Antonio Moreno   |        Owner         |        Mataderos  2312        | México D.F. |  None  |    05023    |  Mexico |  (5) 555-3932  |      None      |
|    AROUT    |          Around the Horn           |    Thomas Hardy    | Sales Representative |        120 Hanover Sq.        |    London   |  None  |   WA1 1DP   |    UK   | (171) 555-7788 | (171) 555-6750 |
|    BERGS    |         Berglunds snabbköp         | Christina Berglund | Order Administrator  |        Berguvsvägen  8        |    Luleå    |  None  |   S-958 22  |  Sweden | 0921-12 34 65  | 0921-12 34 67  |
+-------------+------------------------------------+--------------------+----------------------+-------------------------------+-------------+--------+-------------+---------+----------------+----------------+

### Creating Products and Orders Table

To make later analysis easier, will create views along the way of some tables that we will be using multiple times.

In [51]:
%%sql
CREATE VIEW Products_and_Orders AS
SELECT o.customer_id,
       o.order_id,
       p.product_id,
       p.product_name,
       o.order_date,
       od.unit_price,
       od.quantity, 
       od.discount
  FROM orders o
  JOIN order_details od ON o.order_id = od.order_id
  JOIN products p ON od.product_id = p.product_id;

Running query in 'postgresql://postgres:***@localhost/northwind'

++
||
++
++

## Employee Sales Ranking

We will first examine employee performance with regards to sales made for the company and find the highest and lowest performing employees in order to boost productivity.

#### Ranking employees by total sales

In [64]:
%%sql
WITH Employee_Sales AS
(
    SELECT e.first_name || ' ' || e.last_name AS employee_name,
           ROUND(CAST(SUM(unit_price * quantity * (1 - discount)) AS numeric), 2) AS total_sales,
           CASE
               WHEN SUM(unit_price * quantity * (1-discount)) >
                    AVG(SUM(unit_price * quantity * (1-discount))) OVER () THEN 'Above Average'
               ELSE 'Below Average'
            END AS value_category
      FROM Orders AS o
      JOIN Order_Details AS od ON o.order_id = od.order_id
      JOIN Employees AS e ON o.employee_id = e.employee_id
     GROUP BY e.employee_id
)
SELECT employee_name, total_sales,
       RANK() OVER (ORDER BY total_sales DESC) AS sales_rank,
       ROUND(AVG(total_sales) OVER(), 2) AS average_total_sales, value_category 
  FROM Employee_Sales;

Running query in 'postgresql://postgres:***@localhost/northwind'

9 rows affected.

+------------------+-------------+------------+---------------------+----------------+
|  employee_name   | total_sales | sales_rank | average_total_sales | value_category |
+------------------+-------------+------------+---------------------+----------------+
| Margaret Peacock |  232890.85  |     1      |      140643.67      | Above Average  |
| Janet Leverling  |  202812.84  |     2      |      140643.67      | Above Average  |
|  Nancy Davolio   |  192107.60  |     3      |      140643.67      | Above Average  |
|  Andrew Fuller   |  166537.76  |     4      |      140643.67      | Above Average  |
|  Laura Callahan  |  126862.28  |     5      |      140643.67      | Below Average  |
|   Robert King    |  124568.23  |     6      |      140643.67      | Below Average  |
|  Anne Dodsworth  |   77308.07  |     7      |      140643.67      | Below Average  |
|  Michael Suyama  |   73913.13  |     8      |      140643.67      | Below Average  |
| Steven Buchanan  |   68792.28  |     9      |      140643.67      | Below Average  |
+------------------+-------------+------------+---------------------+----------------+

Four employees have above average sales and should be rewarded, with special considertion given to the highest performer Margaret Peacock.
Five employees have below average sales, so the reasons for their poor performance could be looked into and if needed extra training and resources could be provided to help them improve.

##### Ranking employees by average sale value per order

In [107]:
%%sql
SELECT e.first_name || ' ' || e.last_name AS employee_name, 
       AVG(quantity * unit_price * (1 - discount)) as avg_sales
  FROM orders AS o
  JOIN Order_Details AS od ON o.order_id = od.order_id
  JOIN Employees AS e ON o.employee_id = e.employee_id
 GROUP BY employee_name
 ORDER BY avg_sales DESC


Running query in 'postgresql://postgres:***@localhost/northwind'

9 rows affected.

+------------------+-------------------+
|  employee_name   |     avg_sales     |
+------------------+-------------------+
|  Anne Dodsworth  | 722.5052919467573 |
|   Robert King    | 707.7740616536112 |
|  Andrew Fuller   | 691.0280302382771 |
| Janet Leverling  | 631.8157083734419 |
| Steven Buchanan  | 587.9682255893127 |
|  Nancy Davolio   | 556.8336366866436 |
| Margaret Peacock | 554.5020142429466 |
|  Laura Callahan  |  487.93183645337  |
|  Michael Suyama  | 439.9591034403008 |
+------------------+-------------------+

Anne Dodsworth has the highest average value sale per order, but only ranks 7th in total sales value. Michael Suyama has the lowest average sale value and is also the 2nd lowest performer by total sales value. None of the top 3 performing employees by total sales value rank in the top 3 for average sales value.

##### The Highest Sales recorded and the employee that recorded the sale

In [83]:
%%sql
SELECT e.first_name || ' ' || e.last_name AS employee_name, 
       ROUND(CAST(quantity * unit_price * (1 - discount) AS numeric), 2) as sales
  FROM orders AS o
  JOIN Order_Details AS od ON o.order_id = od.order_id
  JOIN Employees AS e ON o.employee_id = e.employee_id
 ORDER BY sales DESC
 LIMIT 20


Running query in 'postgresql://postgres:***@localhost/northwind'

20 rows affected.

+------------------+----------+
|  employee_name   |  sales   |
+------------------+----------+
|  Nancy Davolio   | 15810.00 |
|  Andrew Fuller   | 15019.50 |
|  Anne Dodsworth  | 10540.00 |
| Margaret Peacock | 10540.00 |
| Janet Leverling  | 9903.20  |
|   Robert King    | 8432.00  |
|   Robert King    | 8263.36  |
| Janet Leverling  | 7905.00  |
| Janet Leverling  | 7905.00  |
| Margaret Peacock | 7509.75  |
|  Andrew Fuller   | 6587.50  |
| Steven Buchanan  | 6324.00  |
| Janet Leverling  | 6324.00  |
|  Anne Dodsworth  | 6050.00  |
|  Nancy Davolio   | 6042.00  |
|  Andrew Fuller   | 5570.55  |
|   Robert King    | 5570.55  |
|  Andrew Fuller   | 5268.00  |
|  Andrew Fuller   | 4951.60  |
|   Robert King    | 4642.13  |
+------------------+----------+

Andrew Fuller was responsible for 5 of the top 20 sales by order value, whilst Robert King and Janet Leverling were both responsible for 4 of the top 20. This is not surprising as these 3 employees rank in the top 4 for average sale value.

##### Total quantity of sales recorded by employees

In [84]:
%%sql
SELECT e.first_name || ' ' || e.last_name AS employee_name, COUNT(quantity * unit_price * (1 - discount)) as sales
  FROM orders AS o
  JOIN Order_Details AS od ON o.order_id = od.order_id
  JOIN Employees AS e ON o.employee_id = e.employee_id
 GROUP BY e.employee_id
 ORDER BY sales DESC


Running query in 'postgresql://postgres:***@localhost/northwind'

9 rows affected.

+------------------+-------+
|  employee_name   | sales |
+------------------+-------+
| Margaret Peacock |  420  |
|  Nancy Davolio   |  345  |
| Janet Leverling  |  321  |
|  Laura Callahan  |  260  |
|  Andrew Fuller   |  241  |
|   Robert King    |  176  |
|  Michael Suyama  |  168  |
| Steven Buchanan  |  117  |
|  Anne Dodsworth  |  107  |
+------------------+-------+

Although Margaret Peacock ranked only 7th for average sale value, she has made the most sales by a considerable margin. Anne Dodsworth has made the least sales, but ranks top in average sale value. Michael Suyama has made the 3rd least sales and has the lowest average sale value, so his performance should be looked into.

Margaret Peacock, Nancy Davolio and Janet Leverling rank in the top 3 in the total sales quantity table and also the total sales value table, suggesting that quantity of sales made is more important than average value per sale. This is echoed by Anne Dodsworth ranking 7th in total sales despite having the highest average sales value, due to the fact that she has made the least amount of sales. 

## Seasonal Sales Analysis

In [68]:
%config SqlMagic.named_parameters="enabled"

###### Creating a view for monthly sales

In [100]:
%%sql 
CREATE VIEW Monthly_Sales AS 
 SELECT DATE_TRUNC('month', order_date)::DATE AS month, 
        SUM(unit_price * quantity * (1 - discount)) AS monthly_sales
   FROM Orders AS o 
   JOIN Order_Details AS od ON o.order_ID = od.order_ID
  GROUP BY DATE_TRUNC('month', order_date)

Running query in 'postgresql://postgres:***@localhost/northwind'

++
||
++
++

### Employees' Seasonal Sales Growth Rate

##### Most recent 5 orders by order date

In [28]:
%%sql
SELECT *
FROM orders
ORDER BY order_date DESC
LIMIT 5;

Running query in 'postgresql://postgres:***@localhost/northwind'

5 rows affected.

+----------+-------------+-------------+------------+---------------+--------------+----------+---------+----------------------------+--------------------------+-------------+-------------+------------------+--------------+
| order_id | customer_id | employee_id | order_date | required_date | shipped_date | ship_via | freight |         ship_name          |       ship_address       |  ship_city  | ship_region | ship_postal_code | ship_country |
+----------+-------------+-------------+------------+---------------+--------------+----------+---------+----------------------------+--------------------------+-------------+-------------+------------------+--------------+
|  11076   |    BONAP    |      4      | 1998-05-06 |   1998-06-03  |     None     |    2     |  38.28  |          Bon app'          |   12, rue des Bouchers   |  Marseille  |     None    |      13008       |    France    |
|  11077   |    RATTC    |      1      | 1998-05-06 |   1998-06-03  |     None     |    2     |   8.53  | Rattlesnake Canyon Grocery |     2817 Milton Dr.      | Albuquerque |      NM     |      87110       |     USA      |
|  11074   |    SIMOB    |      7      | 1998-05-06 |   1998-06-03  |     None     |    2     |  18.44  |       Simons bistro        |       Vinbæltet 34       |  Kobenhavn  |     None    |       1734       |   Denmark    |
|  11075   |    RICSU    |      8      | 1998-05-06 |   1998-06-03  |     None     |    2     |   6.19  |     Richter Supermarkt     |       Starenweg 5        |    Genève   |     None    |       1204       | Switzerland  |
|  11073   |    PERIC    |      2      | 1998-05-05 |   1998-06-02  |     None     |    2     |  24.95  | Pericles Comidas clásicas  | Calle Dr. Jorge Cash 321 | México D.F. |     None    |      05033       |    Mexico    |
+----------+-------------+-------------+------------+---------------+--------------+----------+---------+----------------------------+--------------------------+-------------+-------------+------------------+--------------+

This database only has records for sales up to the 6th day of the most recent month recorded. We will therefore omit the May 1998 information from our analysis as otherwise this would skew sales values. 

In [87]:
%%sql
WITH Employee_Sales AS
(
    SELECT e.first_name || ' ' || e.last_name AS employee_name,
           ROUND(CAST(SUM(unit_price * quantity * (1 - discount)) AS numeric), 2) AS current_month_sales,
           DATE_TRUNC('month', order_date)::DATE AS month
      FROM Orders AS o
      JOIN Order_Details AS od ON o.order_id = od.order_id
      JOIN Employees AS e ON o.employee_id = e.employee_id
     GROUP BY e.employee_id, month
)
SELECT month, employee_name, current_month_sales,
       ROUND(AVG(current_month_sales) OVER (PARTITION BY employee_name ORDER BY month
                                    ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW), 2) AS monthly_average,
       ROUND(AVG(current_month_sales) OVER (PARTITION BY employee_name ORDER BY month 
                                    ROWS BETWEEN 2 PRECEDING AND CURRENT ROW), 2) AS three_month_average,
       ROUND(AVG(current_month_sales) OVER (PARTITION BY employee_name ORDER BY month 
                                    ROWS BETWEEN 11 PRECEDING AND CURRENT ROW), 2) AS twelve_month_average
  FROM Employee_Sales
 WHERE month NOT IN ('1998-05-01')
 ORDER BY month DESC, monthly_average DESC
 LIMIT 9



Running query in 'postgresql://postgres:***@localhost/northwind'

9 rows affected.

+------------+------------------+---------------------+-----------------+---------------------+----------------------+
|   month    |  employee_name   | current_month_sales | monthly_average | three_month_average | twelve_month_average |
+------------+------------------+---------------------+-----------------+---------------------+----------------------+
| 1998-04-01 | Margaret Peacock |       9937.71       |     10312.73    |       9626.41       |       10197.54       |
| 1998-04-01 | Janet Leverling  |       12957.35      |     9218.77     |       16952.57      |       12128.61       |
| 1998-04-01 |  Nancy Davolio   |       12587.22      |     8435.92     |       16187.39      |       11265.30       |
| 1998-04-01 |  Andrew Fuller   |       30990.28      |     7482.17     |       22685.16      |       10186.94       |
| 1998-04-01 |   Robert King    |       28590.56      |     6170.37     |       13698.01      |       7020.66        |
| 1998-04-01 |  Laura Callahan  |       13777.10      |     5643.08     |       11537.08      |       6868.56        |
| 1998-04-01 |  Anne Dodsworth  |       9501.50       |     4294.89     |       11916.61      |       5364.73        |
| 1998-04-01 | Steven Buchanan  |        210.00       |     3821.79     |       2663.03       |       3990.66        |
| 1998-04-01 |  Michael Suyama  |       5246.95       |     3519.67     |       4089.78       |       3648.13        |
+------------+------------------+---------------------+-----------------+---------------------+----------------------+

Although Margaret Peacock is the highest performing employee, her average sales by value in the past 3 months is below her monthly average and is lower than 6 other employees. 

Janet Leverling has the highest average based on the past year. She also has the 2nd highest monthly average and the 2nd highest average in the past 3 months.

Andrew Fuller and Robert King both have very high sale values this month. This is not that far off Andrew Fuller's 3 month average suggesting his performance has been strong in recent months, whilst the value for Robert King is much higher than his average in the past 3 months, showing he has had an unsually strong month.

The lowest performing employee Michael Suyama has a sales value this month which is higher than his averages, suggesting his performace is improving.

Steven Buchanan only has a sales value of 210 for April 1998, which is much below his average values so we will examine how many sales he made last month to see why this is.

In general, 7 of the 9 employees have a higher 3 month average than their overall monthly average, which is likely a result of the company becoming more popular and taking more orders as they grow over time.

###### Verifying Steven Buchanan's recent performance 

In [114]:
%%sql
SELECT e.first_name || ' ' || e.last_name AS employee_name,
           ROUND(CAST(COUNT(unit_price * quantity * (1 - discount)) AS numeric), 2) AS Number_sales_this_month,
           DATE_TRUNC('month', order_date)::DATE AS month
      FROM Orders AS o
      JOIN Order_Details AS od ON o.order_id = od.order_id
      JOIN Employees AS e ON o.employee_id = e.employee_id
     WHERE e.first_name || ' ' || e.last_name IN ('Steven Buchanan')
        GROUP BY e.employee_id, month
        ORDER BY month DESC
    LIMIT 1;

Running query in 'postgresql://postgres:***@localhost/northwind'

1 rows affected.

+-----------------+-------------------------+------------+
|  employee_name  | number_sales_this_month |   month    |
+-----------------+-------------------------+------------+
| Steven Buchanan |           1.00          | 1998-04-01 |
+-----------------+-------------------------+------------+

We can see that Steven Buchanan only made 1 sale in April 1998, so this explains why his sales value for this month was so low. This could be down to external factors such as time off rather than poor performance.

### Sales growth rate over the past year

In [103]:
%%sql
WITH Previous_Month_Sales AS
(
    SELECT month, monthly_sales,
           LAG(monthly_sales) OVER (ORDER BY month) AS lagged_sales
      FROM Monthly_Sales
)
SELECT month,
       ROUND(CAST(monthly_sales AS numeric), 2) AS monthly_sales,
       ROUND(CAST(monthly_sales - lagged_sales AS numeric), 2) AS difference,
       ROUND(CAST(SUM(monthly_sales) OVER (ORDER BY month) AS numeric), 2) AS running_total,
       CASE 
           WHEN monthly_sales - lagged_sales > 0 THEN 'Increase'
           WHEN monthly_sales - lagged_sales < 0 THEN 'Decrease'
           ELSE 'Decrease' # setting the first value as non-zero using growth rate value 
        END AS revenue_change,
        ROUND(CAST(((monthly_sales - lagged_sales) / lagged_sales) * 100
                   AS numeric), 2) AS growth_rate
FROM (SELECT * 
      FROM Previous_Month_Sales
      WHERE month NOT IN ('1998-05-01')
      ORDER BY month DESC
      LIMIT 12) AS last_12_months
ORDER BY month; 

Running query in 'postgresql://postgres:***@localhost/northwind'

12 rows affected.

+------------+---------------+------------+---------------+----------------+-------------+
|   month    | monthly_sales | difference | running_total | revenue_change | growth_rate |
+------------+---------------+------------+---------------+----------------+-------------+
| 1997-05-01 |    53781.29   |   748.34   |    53781.29   |    Increase    |     1.41    |
| 1997-06-01 |    36362.80   | -17418.49  |    90144.09   |    Decrease    |    -32.39   |
| 1997-07-01 |    51020.86   |  14658.06  |   141164.95   |    Increase    |    40.31    |
| 1997-08-01 |    47287.67   |  -3733.19  |   188452.62   |    Decrease    |    -7.32    |
| 1997-09-01 |    55629.24   |  8341.57   |   244081.86   |    Increase    |    17.64    |
| 1997-10-01 |    66749.23   |  11119.98  |   310831.09   |    Increase    |    19.99    |
| 1997-11-01 |    43533.81   | -23215.42  |   354364.90   |    Decrease    |    -34.78   |
| 1997-12-01 |    71398.43   |  27864.62  |   425763.32   |    Increase    |    64.01    |
| 1998-01-01 |    94222.11   |  22823.68  |   519985.43   |    Increase    |    31.97    |
| 1998-02-01 |    99415.29   |  5193.18   |   619400.72   |    Increase    |     5.51    |
| 1998-03-01 |   104854.16   |  5438.87   |   724254.88   |    Increase    |     5.47    |
| 1998-04-01 |   123798.68   |  18944.53  |   848053.56   |    Increase    |    18.07    |
+------------+---------------+------------+---------------+----------------+-------------+

The past year has seen in general good sales performance, with sales increasing month to month 9 out of 12 times. The past 5 months have all seen growth in sales each month, with the largest increase in sales coming between November and December 1997, most likely due to the Christmas holidays. The largest decrease came just before this, with a 35% decrease in sales between October and November 1997.

### Yearly Sales Growth Rate

In [110]:
%%sql
WITH Previous_Year_Sales AS
(
    SELECT month, monthly_sales,
           LAG(monthly_sales, 12) OVER (ORDER BY month) AS lagged_sales
      FROM Monthly_Sales
)
SELECT month,
       ROUND(CAST(monthly_sales AS numeric), 2) AS monthly_sales,
       ROUND(CAST(lagged_sales AS numeric), 2) AS previous_year_sales,
       ROUND(CAST(monthly_sales - lagged_sales AS numeric), 2) AS difference,
       CASE 
           WHEN monthly_sales - lagged_sales > 0 THEN 'Increase'
           WHEN monthly_sales - lagged_sales < 0 THEN 'Decrease'
           ELSE 'N\A' 
        END AS annual_revenue_change,
        ROUND(CAST(((monthly_sales - lagged_sales) / lagged_sales) * 100
                   AS numeric), 2) AS annual_growth_rate
FROM (SELECT * 
      FROM Previous_Year_Sales
      WHERE month NOT IN ('1998-05-01')
      ORDER BY month DESC LIMIT 12) AS last_12_months
ORDER BY month;

Running query in 'postgresql://postgres:***@localhost/northwind'

12 rows affected.

+------------+---------------+---------------------+------------+-----------------------+--------------------+
|   month    | monthly_sales | previous_year_sales | difference | annual_revenue_change | annual_growth_rate |
+------------+---------------+---------------------+------------+-----------------------+--------------------+
| 1997-05-01 |    53781.29   |         None        |    None    |          N\A          |        None        |
| 1997-06-01 |    36362.80   |         None        |    None    |          N\A          |        None        |
| 1997-07-01 |    51020.86   |       27861.90      |  23158.96  |        Increase       |       83.12        |
| 1997-08-01 |    47287.67   |       25485.28      |  21802.39  |        Increase       |       85.55        |
| 1997-09-01 |    55629.24   |       26381.40      |  29247.84  |        Increase       |       110.87       |
| 1997-10-01 |    66749.23   |       37515.72      |  29233.50  |        Increase       |       77.92        |
| 1997-11-01 |    43533.81   |       45600.05      |  -2066.24  |        Decrease       |       -4.53        |
| 1997-12-01 |    71398.43   |       45239.63      |  26158.80  |        Increase       |       57.82        |
| 1998-01-01 |    94222.11   |       61258.07      |  32964.04  |        Increase       |       53.81        |
| 1998-02-01 |    99415.29   |       38483.63      |  60931.65  |        Increase       |       158.33       |
| 1998-03-01 |   104854.16   |       38547.22      |  66306.93  |        Increase       |       172.01       |
| 1998-04-01 |   123798.68   |       53032.95      |  70765.73  |        Increase       |       133.44       |
+------------+---------------+---------------------+------------+-----------------------+--------------------+

The total sales have increased significantly in each month compared with the same month a year before. March 1998 saw the biggest increase from a year before in March 1997, with a 172% increase in sales. The other months all showed a greater than 50% growth rate, aside from a small decrease in November. This highlights the strongth growth of the company with regards to sales which was equally echoed in improved employee performance over time.

The N/A values can be explained by no sales records for May and June the previous year in 1996.

## High-Value Customers

We want to identify the customers with above average order values, so that we can target these customers with special offers in an effort to increase sales and improve customer retention.

#### Largest Orders by Value

In [30]:
%%sql 
SELECT customer_id, order_id, product_name, 
       ROUND(CAST(order_value AS numeric), 2) AS order_value
  FROM (
          SELECT customer_id,
                 order_id,
                 product_name,
                 SUM(unit_price * quantity * (1-discount)) AS order_value
            FROM Products_and_Orders
           GROUP BY customer_ID, order_ID, product_name
       ) AS Customer_Order_Values
 WHERE order_value > 5000
 ORDER BY order_value DESC;
        
        

Running query in 'postgresql://postgres:***@localhost/northwind'

18 rows affected.

+-------------+----------+-------------------------+-------------+
| customer_id | order_id |       product_name      | order_value |
+-------------+----------+-------------------------+-------------+
|    HANAR    |  10981   |      Côte de Blaye      |   15810.00  |
|    QUICK    |  10865   |      Côte de Blaye      |   15019.50  |
|    RATTC    |  10889   |      Côte de Blaye      |   10540.00  |
|    SIMOB    |  10417   |      Côte de Blaye      |   10540.00  |
|    HUNGO    |  10897   | Thüringer Rostbratwurst |   9903.20   |
|    PICCO    |  10353   |      Côte de Blaye      |   8432.00   |
|    MEREP    |  10424   |      Côte de Blaye      |   8263.36   |
|    KOENE    |  10817   |      Côte de Blaye      |   7905.00   |
|    QUICK    |  10540   |      Côte de Blaye      |   7905.00   |
|    GREAL    |  10816   |      Côte de Blaye      |   7509.75   |
|    WHITC    |  11032   |      Côte de Blaye      |   6587.50   |
|    QUEEN    |  10372   |      Côte de Blaye      |   6324.00   |
|    RATTC    |  10479   |      Côte de Blaye      |   6324.00   |
|    ERNSH    |  11017   |   Raclette Courdavault  |   6050.00   |
|    ERNSH    |  10776   |  Manjimup Dried Apples  |   6042.00   |
|    HUNGO    |  10912   | Thüringer Rostbratwurst |   5570.55   |
|    SAVEA    |  11030   | Thüringer Rostbratwurst |   5570.55   |
|    QUICK    |  10515   |    Schoggi Schokolade   |   5268.00   |
+-------------+----------+-------------------------+-------------+

There were 4 orders with a value more than 10,000, with the most valuable order coming from the customer with id HANAR. We see that the majority of these orders comprised of the Côte de Blaye product, emphasising it as a high value product.

### Customers with most Above Average Order Values

In [31]:
%%sql
WITH order_values AS 
(
    SELECT customer_id, product_name, 
       SUM(unit_price * quantity * (1-discount)) AS order_value,
    CASE
           WHEN SUM(unit_price * quantity * (1-discount)) >
                AVG(SUM(unit_price * quantity * (1-discount))) OVER () THEN 'Above Average'
           ELSE 'Below Average'
       END AS value_category
  FROM Products_and_Orders
  GROUP BY customer_id, product_name
)
SELECT customer_id, value_category, COUNT(value_category)
FROM order_values
 GROUP BY customer_id, value_category
HAVING value_category LIKE 'Above Average'
 ORDER BY count DESC
LIMIT 10;

Running query in 'postgresql://postgres:***@localhost/northwind'

10 rows affected.

+-------------+----------------+-------+
| customer_id | value_category | count |
+-------------+----------------+-------+
|    SAVEA    | Above Average  |   40  |
|    ERNSH    | Above Average  |   37  |
|    QUICK    | Above Average  |   33  |
|    FOLKO    | Above Average  |   15  |
|    HUNGO    | Above Average  |   15  |
|    KOENE    | Above Average  |   12  |
|    RATTC    | Above Average  |   12  |
|    SUPRD    | Above Average  |   12  |
|    BONAP    | Above Average  |   11  |
|    HILAA    | Above Average  |   11  |
+-------------+----------------+-------+

The customers Savea, Ernsh and Quick have all made more than 30 orders which were above the average order value, so these are the customers we should target.

#### Ranking Customers by Total Order Value

In [32]:
%%sql
WITH Customer_Purchases AS
(
    SELECT customer_id,
           ROUND(CAST(SUM(unit_price * quantity * (1-discount)) AS numeric), 2) AS order_value
      FROM Products_and_Orders
     GROUP BY customer_id
)
SELECT customer_id, order_value,
       RANK() OVER (ORDER BY order_value DESC) AS order_value_rank
  FROM Customer_Purchases
  GROUP BY customer_id, order_value
  LIMIT 10;

Running query in 'postgresql://postgres:***@localhost/northwind'

10 rows affected.

+-------------+-------------+------------------+
| customer_id | order_value | order_value_rank |
+-------------+-------------+------------------+
|    QUICK    |  110277.31  |        1         |
|    ERNSH    |  104874.98  |        2         |
|    SAVEA    |  104361.95  |        3         |
|    RATTC    |   51097.80  |        4         |
|    HUNGO    |   49979.91  |        5         |
|    HANAR    |   32841.37  |        6         |
|    KOENE    |   30908.38  |        7         |
|    FOLKO    |   29567.56  |        8         |
|    MEREP    |   28872.19  |        9         |
|    WHITC    |   27363.60  |        10        |
+-------------+-------------+------------------+

We again see the same customers in the top 3 places, with each having total order values exceeding 100,000.

#### Top 20% of Customers by Order Quantity

In [33]:
%%sql
WITH customer_quantity AS 
(
    SELECT customer_id, SUM(quantity) AS total_quantity, 
    ROUND(CAST(AVG(quantity) AS numeric), 2) AS average_quantity,
    MIN(quantity) AS min_quantity, MAX(quantity) AS max_quantity
FROM Orders o
JOIN Order_Details od ON o.order_id = od.order_id
GROUP BY customer_id
ORDER BY total_quantity DESC

)
SELECT customer_id, total_quantity, average_quantity, min_quantity, max_quantity
FROM customer_quantity
WHERE total_quantity >= (
    SELECT PERCENTILE_DISC(0.8) WITHIN GROUP (ORDER BY total_quantity)
       AS "80th Percentile"
    FROM customer_quantity
)
ORDER BY total_quantity DESC;

Running query in 'postgresql://postgres:***@localhost/northwind'

18 rows affected.

+-------------+----------------+------------------+--------------+--------------+
| customer_id | total_quantity | average_quantity | min_quantity | max_quantity |
+-------------+----------------+------------------+--------------+--------------+
|    SAVEA    |      4958      |      42.74       |      3       |     120      |
|    ERNSH    |      4543      |      44.54       |      4       |     130      |
|    QUICK    |      3961      |      46.06       |      7       |     120      |
|    HUNGO    |      1684      |      30.62       |      2       |      80      |
|    FRANK    |      1525      |      31.77       |      3       |      70      |
|    RATTC    |      1383      |      19.48       |      1       |      70      |
|    FOLKO    |      1234      |      27.42       |      2       |      60      |
|    HILAA    |      1096      |      24.36       |      2       |      70      |
|    SUPRD    |      1072      |      27.49       |      2       |      60      |
|    WHITC    |      1063      |      26.58       |      2       |      70      |
|    QUEEN    |      1031      |      25.78       |      2       |      70      |
|    BERGS    |      1001      |      19.25       |      2       |      60      |
|    BONAP    |      980       |      22.27       |      3       |      50      |
|    LINOD    |      970       |      27.71       |      4       |      60      |
|    MEREP    |      966       |      30.19       |      3       |      70      |
|    BOTTM    |      956       |      27.31       |      5       |      60      |
|    KOENE    |      903       |      23.15       |      4       |      60      |
|    HANAR    |      839       |      26.22       |      4       |      70      |
+-------------+----------------+------------------+--------------+--------------+

SAVEA, ERNSH and QUICK have made the most orders as well as having the highest order values. The gap between these 3 and the rest of the customers is quite large so these are by far the 3 highest value customers, and we should target them with promotional incentives.

## Best Performing Product Categories

##### Products ranked by percentage of total sales

In [34]:
%%sql
WITH Category_Sales AS
(
    SELECT c.category_id, c.category_name,
           SUM(p.unit_price * quantity * (1 - discount)) AS total_sales
      FROM Categories AS c
      JOIN Products AS p ON c.category_id = p.category_id
      JOIN Order_Details AS od ON p.product_id = od.product_id
     GROUP BY c.category_id
)
SELECT category_id, category_name,
       ROUND(CAST(total_sales / SUM(total_sales) 
                  OVER() * 100 AS numeric), 2) AS sales_percentage
  FROM Category_Sales
 ORDER BY sales_percentage DESC;

Running query in 'postgresql://postgres:***@localhost/northwind'

8 rows affected.

+-------------+----------------+------------------+
| category_id | category_name  | sales_percentage |
+-------------+----------------+------------------+
|      1      |   Beverages    |      21.33       |
|      4      | Dairy Products |      18.56       |
|      3      |  Confections   |      13.29       |
|      6      |  Meat/Poultry  |      12.90       |
|      8      |    Seafood     |      10.20       |
|      2      |   Condiments   |       8.40       |
|      7      |    Produce     |       7.81       |
|      5      | Grains/Cereals |       7.51       |
+-------------+----------------+------------------+

Beverages and dairy products combined account for almost 40% of company sales, so more focus should be put on insuring inventories for these categories are fully stocked. Equally it may be wise to push these products more aggressively in marketing strategies. 

Grains/Cereals ranks lowest and is the least important category with regards to sales.

## Top 3 Products per Category

In [111]:
%%sql 
CREATE VIEW Product_Sales AS 
SELECT c.category_name, 
           p.product_id, p.product_name,
           ROUND(CAST(SUM(p.unit_price * quantity * (1 - discount)) AS numeric), 2) AS total_sales
    FROM Categories AS c
    JOIN Products AS p ON c.category_id = p.category_id
    JOIN Order_Details AS od ON p.product_id = od.product_id
    GROUP BY c.category_name, p.product_id

Running query in 'postgresql://postgres:***@localhost/northwind'

++
||
++
++

In [112]:
%%sql

SELECT category_name, product_ID, product_name,total_sales
FROM (
    SELECT category_name, Product_ID, product_name,total_sales, 
           ROW_NUMBER() OVER (PARTITION BY category_name ORDER BY total_sales DESC) AS row_number
    FROM Product_Sales
) AS ranking
WHERE row_number <= 3;

Running query in 'postgresql://postgres:***@localhost/northwind'

24 rows affected.

+----------------+------------+----------------------------------+-------------+
| category_name  | product_id |           product_name           | total_sales |
+----------------+------------+----------------------------------+-------------+
|   Beverages    |     38     |          Côte de Blaye           |  153897.17  |
|   Beverages    |     43     |           Ipoh Coffee            |   25109.10  |
|   Beverages    |     2      |              Chang               |   17719.40  |
|   Condiments   |     63     |           Vegie-spread           |   18343.62  |
|   Condiments   |     61     |          Sirop d'érable          |   15022.35  |
|   Condiments   |     65     | Louisiana Fiery Hot Pepper Sauce |   14893.93  |
|  Confections   |     62     |          Tarte au sucre          |   50737.09  |
|  Confections   |     20     |      Sir Rodney's Marmalade      |   24199.56  |
|  Confections   |     26     |       Gumbär Gummibärchen        |   21662.69  |
| Dairy Products |     59     |       Raclette Courdavault       |   76683.75  |
| Dairy Products |     60     |        Camembert Pierrot         |   49877.32  |
| Dairy Products |     72     |      Mozzarella di Giovanni      |   27086.58  |
| Grains/Cereals |     56     |      Gnocchi di nonna Alice      |   45351.10  |
| Grains/Cereals |     64     |    Wimmers gute Semmelknödel     |   23487.47  |
| Grains/Cereals |     42     |  Singaporean Hokkien Fried Mee   |   8986.60   |
|  Meat/Poultry  |     29     |     Thüringer Rostbratwurst      |   84783.77  |
|  Meat/Poultry  |     17     |           Alice Mutton           |   35105.85  |
|  Meat/Poultry  |     53     |          Perth Pasties           |   22623.80  |
|    Produce     |     51     |      Manjimup Dried Apples       |   43846.90  |
|    Produce     |     28     |        Rössle Sauerkraut         |   27936.84  |
|    Produce     |     7      | Uncle Bob's Organic Dried Pears  |   22453.50  |
|    Seafood     |     18     |         Carnarvon Tigers         |   30728.12  |
|    Seafood     |     10     |              Ikura               |   21653.50  |
|    Seafood     |     40     |         Boston Crab Meat         |   19055.04  |
+----------------+------------+----------------------------------+-------------+

These are the 3 most valuable products per category and it is important that these are well stocked in order to ensure good sales performance.

### Top 25% of Products by Total Sales 

In [113]:
%%sql

SELECT category_name, product_ID, product_name,total_sales, percent_rank
FROM( 
    SELECT category_name, Product_ID, product_name,total_sales,
    ROUND(CAST(PERCENT_RANK() OVER(ORDER BY total_sales DESC) AS numeric), 2) AS percent_rank
    FROM Product_Sales
) AS percentiles
WHERE percent_rank <=0.25;
    
    


Running query in 'postgresql://postgres:***@localhost/northwind'

20 rows affected.

+----------------+------------+---------------------------------+-------------+--------------+
| category_name  | product_id |           product_name          | total_sales | percent_rank |
+----------------+------------+---------------------------------+-------------+--------------+
|   Beverages    |     38     |          Côte de Blaye          |  153897.17  |     0.00     |
|  Meat/Poultry  |     29     |     Thüringer Rostbratwurst     |   84783.77  |     0.01     |
| Dairy Products |     59     |       Raclette Courdavault      |   76683.75  |     0.03     |
|  Confections   |     62     |          Tarte au sucre         |   50737.09  |     0.04     |
| Dairy Products |     60     |        Camembert Pierrot        |   49877.32  |     0.05     |
| Grains/Cereals |     56     |      Gnocchi di nonna Alice     |   45351.10  |     0.07     |
|    Produce     |     51     |      Manjimup Dried Apples      |   43846.90  |     0.08     |
|  Meat/Poultry  |     17     |           Alice Mutton          |   35105.85  |     0.09     |
|    Seafood     |     18     |         Carnarvon Tigers        |   30728.12  |     0.11     |
|    Produce     |     28     |        Rössle Sauerkraut        |   27936.84  |     0.12     |
| Dairy Products |     72     |      Mozzarella di Giovanni     |   27086.58  |     0.13     |
|   Beverages    |     43     |           Ipoh Coffee           |   25109.10  |     0.14     |
|  Confections   |     20     |      Sir Rodney's Marmalade     |   24199.56  |     0.16     |
| Grains/Cereals |     64     |    Wimmers gute Semmelknödel    |   23487.47  |     0.17     |
| Dairy Products |     69     |         Gudbrandsdalsost        |   23311.80  |     0.18     |
|  Meat/Poultry  |     53     |          Perth Pasties          |   22623.80  |     0.20     |
|    Produce     |     7      | Uncle Bob's Organic Dried Pears |   22453.50  |     0.21     |
|  Confections   |     26     |       Gumbär Gummibärchen       |   21662.69  |     0.22     |
|    Seafood     |     10     |              Ikura              |   21653.50  |     0.24     |
| Dairy Products |     71     |           Flotemysost           |   21339.82  |     0.25     |
+----------------+------------+---------------------------------+-------------+--------------+

We confirm what we saw earlier that Côte de Blaye is the most valuable product, with its total sales exceeding second place by almost 70,000. This product is followed by Thüringer Rostbratwurst and Raclette Courdavault as the most valuable products, with Tarte au sucre and Camembert Pierrot rounding of the top 5% of products by total sales.

There is a wide variety of categories in the top 25% of products and it is not dominated by any one in particular.

## Conclusion

With these findings the company can continue to make profits and retain key customers and employees.